File: app/src/main/java/com/nvvi9/giphyme/utils/MultiSpringAddListener.java
package com.nvvi9.giphyme.utils;

import androidx.dynamicanimation.animation.DynamicAnimation;
import androidx.dynamicanimation.animation.SpringAnimation;

import java.util.ArrayList;
import java.util.List;
import java.util.function.Consumer;

public class MultiSpringAddListener {

    private final List<DynamicAnimation.OnAnimationEndListener> listeners;
    private boolean wasCancelled = false;

    public MultiSpringAddListener(Consumer<Boolean> onEnd, SpringAnimation... springs) {
        listeners = new ArrayList<>(springs.length);
        for (SpringAnimation springAnimation : springs) {
            DynamicAnimation.OnAnimationEndListener onAnimationEndListener = new DynamicAnimation.OnAnimationEndListener() {
                @Override
                public void onAnimationEnd(DynamicAnimation animation, boolean canceled, float value, float velocity) {
                    if (animation != null) {
                        animation.removeEndListener(this);
                    }
                    wasCancelled = wasCancelled || canceled;
                    listeners.remove(this);
                    if (listeners.isEmpty()) {
                        onEnd.accept(wasCancelled);
                    }
                }
            };

            springAnimation.addEndListener(onAnimationEndListener);
            listeners.add(onAnimationEndListener);
        }
    }
}

File: app/src/main/java/com/nvvi9/giphyme/utils/MultiSpringAddListener.kt
package com.nvvi9.giphyme.utils

import androidx.dynamicanimation.animation.DynamicAnimation
import androidx.dynamicanimation.animation.DynamicAnimation.OnAnimationEndListener
import androidx.dynamicanimation.animation.SpringAnimation

class MultiSpringAddListener(onEnd: (Boolean) -> Unit, vararg springs: SpringAnimation) {
    private val listeners: List<OnAnimationEndListener>
    private var wasCancelled = false

    init {
        listeners = mutableListOf()
        for (springAnimation in springs) {
            val onAnimationEndListener: OnAnimationEndListener = object : OnAnimationEndListener {
                override fun onAnimationEnd(
                    animation: DynamicAnimation<*>?,
                    canceled: Boolean,
                    value: Float,
                    velocity: Float
                ) {
                    animation?.removeEndListener(this)
                    wasCancelled = wasCancelled || canceled
                    listeners.remove(this)
                    if (listeners.isEmpty()) {
                        onEnd(wasCancelled)
                    }
                }
            }
            springAnimation.addEndListener(onAnimationEndListener)
            listeners.add(onAnimationEndListener)
        }
    }
}